### PR TITLE
Plfm 3364 d

### PR DESCRIPTION
**Pull request by john-hill**

### CODE REVIEW

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+
+	static private Logger log = LogManager
+			.getLogger(BulkFileDownloadWorker.class);
+
+	/**
+	 * The maximum total size in bytes of generated zip files.
+	 */
+	public static final long MAX_TOTAL_FILE_SIZE_BYTES = 1024 * 1024 * 1024; // 1 GB.
+
+	@Autowired
```
Comment (brucehoff) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java:67]: why is one manager declared private and the other two not

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+				summary.setFailureMessage(e.getMessage());
+				summary.setFailureCode(FileDownloadCode.NOT_FOUND);
+			} catch (Exception e) {
+				// all unknown errors.
+				summary.setStatus(FileDownloadStatus.FAILURE);
+				summary.setFailureMessage(e.getMessage());
+				summary.setFailureCode(FileDownloadCode.UNKNOWN_ERROR);
+				log.error("Failed on: " + fhas.getAssociation(), e);
+			}
+			totalProgress++;
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java:207]: increment current progress

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+	 * @param zipOut
+	 */
+	public List<FileDownloadSummary> addFilesToZip(
+			ProgressCallback<Message> progressCallback, Message message,
+			List<FileHandleAssociationAuthorizationStatus> authResults,
+			File tempResultFile, ZipOutputStream zipOut,
+			AsynchronousJobStatus status,
+			Set<String> fileIdsInZip) {
+		long currentProgress = 0L;
+		long totalProgress = (long) authResults.size();
```
Comment (brucehoff) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java:170]: declare totalProgress to be final

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+			throw new BulkFileException(FILE_EXCEEDS_THE_MAXIMUM_SIZE_LIMIT,
+					FileDownloadCode.EXCEEDS_SIZE_LIMIT);
+		}
+		// This file will be downloaded to this temp.
+		File downloadTemp = bulkDownloadDao.downloadToTempFile(s3Handle);
+		try {
+			// The entry name is the path plus file name.
+			String zipEntryName = createZipEntryName(s3Handle.getFileName(),
+					Long.parseLong(s3Handle.getId()));
+			// write the file to the zip.
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkFileDownloadWorker.java:254]: FileSummary included the zipEntryName.

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkDownloadDaoImp.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+import org.sagebionetworks.repo.model.file.FileHandleAssociation;
+import org.sagebionetworks.repo.model.file.S3FileHandle;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.amazonaws.event.ProgressListener;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+
+public class BulkDownloadDaoImp implements BulkDownloadDao{
+
```
Comment (john-hill) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkDownloadDaoImp.java:28]: Call it a manager.

File: services/workers/src/test/java/org/sagebionetworks/file/worker/BulkFileDownloadWorkerTest.java [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+		// progress should be made for each file.
+		verify(mockProgress, times(2)).progressMade(message);
+		verify(mockAsynchJobStatusManager, times(2)).updateJobProgress(
+				anyString(), anyLong(), anyLong(), anyString());
+		
+		verifyAllStreamsClosedAndFilesDeleted();
+		// The zip should get uploaded
+		verify(mockBulkDownloadDao, times(1)).multipartUploadLocalFile(
+				any(UserInfo.class), any(File.class), anyString(),
+				any(ProgressListener.class));
```
Comment (john-hill) [inline services/workers/src/test/java/org/sagebionetworks/file/worker/BulkFileDownloadWorkerTest.java:294]: add to testHappy()

File: services/workers/src/test/java/org/sagebionetworks/file/worker/BulkFileDownloadWorkerTest.java [hunk 7]
```diff
@@ -0,0 +1,509 @@
...
+	 * The zip has grown past the maximum size.
+	 * @throws Exception
+	 */
+	@Test
+	public void testRunZipFull() throws Exception {
+		// temp file for the zip should be created
+		File mockZip = Mockito.mock(File.class);
+		when(mockBulkDownloadDao.createTempFile(anyString(), anyString())).thenReturn(mockZip);
+		when(mockZip.length()).thenReturn(BulkFileDownloadWorker.MAX_TOTAL_FILE_SIZE_BYTES+1);
+		fileHandle1.setContentSize(BulkFileDownloadWorker.MAX_TOTAL_FILE_SIZE_BYTES+1);
```
Comment (john-hill) [inline services/workers/src/test/java/org/sagebionetworks/file/worker/BulkFileDownloadWorkerTest.java:406]: don't set the file size to be so big.

File: client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java [hunk 8]
```diff
@@ -605,6 +609,24 @@ public void setFileEndpoint(String fileEndpoint) {
 	public String getFileEndpoint() {
 		return this.fileEndpoint;
 	}
+	
+	/*
+	 * (non-Javadoc)
+	 * @see org.sagebionetworks.client.SynapseClient#getEndpointForType(org.sagebionetworks.client.RestEndpointType)
+	 */
+	@Override
+	public String getEndpointForType(RestEndpointType type){
```
Comment (john-hill) [inline client/synapseJavaClient/src/main/java/org/sagebionetworks/client/SynapseClientImpl.java:618]: make protected

File: lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/file/FileDownloadSummary.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
+{
+	"description": "Summary of a requested file download.",
+	"properties": {
+		"fileHandleId": {
+			"type": "string",
+			"description": "The FileHandle.id of a requested file."
```
Comment (brucehoff) [inline lib/lib-auto-generated/src/main/resources/schema/org/sagebionetworks/repo/model/file/FileDownloadSummary.json:6]: typo (FileHandle.id -> file handle id)

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthorizationManagerImpl.java [hunk 10]
```diff
@@ -4,11 +4,15 @@
 import java.util.Collection;
 import java.util.HashSet;
 import java.util.List;
+import java.util.Map;
 import java.util.Map.Entry;
 import java.util.Set;
 
 import org.sagebionetworks.evaluation.manager.EvaluationPermissionsManager;
 import org.sagebionetworks.reflection.model.PaginatedResults;
+import org.sagebionetworks.repo.manager.file.FileAssociateObject;
```
Comment (john-hill) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/AuthorizationManagerImpl.java:13]: clean up imports

File: services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileAssociateObject.java [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+				+ ((objectType == null) ? 0 : objectType.hashCode());
+		return result;
+	}
+	@Override
+	public boolean equals(Object obj) {
+		if (this == obj)
+			return true;
+		if (obj == null)
+			return false;
+		if (getClass() != obj.getClass())
```
Comment (DwayneJengSage) [inline services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/file/FileAssociateObject.java:49]: http://www.jqno.nl/equalsverifier/ can be used to automatically test equals() methods

File: services/workers/src/main/java/org/sagebionetworks/file/worker/BulkDownloadDaoImp.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+import org.sagebionetworks.repo.model.file.FileHandle;
+import org.sagebionetworks.repo.model.file.FileHandleAssociation;
+import org.sagebionetworks.repo.model.file.S3FileHandle;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import com.amazonaws.event.ProgressListener;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+
+public class BulkDownloadDaoImp implements BulkDownloadDao{
```
Comment (DwayneJengSage) [inline services/workers/src/main/java/org/sagebionetworks/file/worker/BulkDownloadDaoImp.java:27]: Imp or Impl?

File: services/workers/src/main/resources/file-bulk-download-spb.xml [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+								</property>
+							</bean>
+						</constructor-arg>
+					</bean>
+				</property>
+				<property name="targetMethod" value="run" />
+				<property name="concurrent" value="false" />
+			</bean>
+		</property>
+		<!-- We stager the start delay of each trigger to spread out the timing -->
```
Comment (brucehoff) [inline services/workers/src/main/resources/file-bulk-download-spb.xml:48]: 'stager' -> 'stagger'

**Pull request merged**